gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,446 @@
+        """
+        self.vector_size = int(vector_size)
+        self.workers = int(workers)
+        self.epochs = epochs
+        self.train_count = 0
+        self.total_train_time = 0
+        self.batch_words = batch_words
+        self.model_trimmed_post_training = False
+        self.callbacks = callbacks
+
+    def build_vocab(self, data_iterable, update=False, progress_per=10000, **kwargs):
+        """Scan through all the data and create/update vocabulary.
+        Should also initialize/reset/update vectors for new vocab entities.
+        """
+        self.vocabulary.scan_vocab(data_iterable, progress_per=progress_per, **kwargs)
+        self.vocabulary.prepare_vocab(update=update, **kwargs)
+        self.trainables.prepare_weights(update=update, vocabulary=self.vocabulary)
+        self._set_keyedvectors()
+
+    def _get_job_params(self):
+        """Return job parameters required for each batch"""
+        raise NotImplementedError
+
+    def _set_train_params(self, **kwargs):
+        """Set model parameters required for training"""
+        raise NotImplementedError
+
+    def _update_job_params(self, job_params, epoch_progress, cur_epoch):
+        """Return updated job parameters based on the epoch_progress and cur_epoch"""
+        raise NotImplementedError
+
+    def _get_thread_working_mem(self):
+        """Return private working memory per thread"""
+        raise NotImplementedError
+
+    def _raw_word_count(self, job):
+        """Return the number of words in a given job."""
+        raise NotImplementedError
+
+    def _clear_post_train(self):
+        """Resets certain properties of the model post training. eg. `kv.vectors_norm`"""
+        raise NotImplementedError
+
+    def _do_train_job(self, batch_data, job_parameters, thread_private_mem):
+        """Train a single batch. Return 2-tuple `(effective word count, total word count)`."""
+        raise NotImplementedError
+
+    def _check_training_sanity(self, epochs=None, total_examples=None, total_words=None, **kwargs):
+        """Check that the training parameters provided make sense. e.g. raise error if `epochs` not provided"""
+        raise NotImplementedError
+
+    def _set_keyedvectors(self):
+        """Point `keyedvectors` attributes to corresponding `trainables`"""
+        self.kv.vectors = self.trainables.vectors
+        self.kv.vector_size = self.trainables.vector_size
+        self.kv.vocab = self.vocabulary.vocab
+        self.kv.index2entity = self.vocabulary.index2word
+
+    def _worker_loop(self, job_queue, progress_queue):
+        """Train the model, lifting lists of data from the job_queue."""
+        thread_private_mem = self._get_thread_working_mem()
+        jobs_processed = 0
+        while True:
+            job = job_queue.get()
+            if job is None:
+                progress_queue.put(None)
+                break  # no more jobs => quit this worker
+            sentences, job_parameters = job
```

Comment:
It would be nice to unify variable naming. We have `sentences` here, `data_iterable` or `batch_data` elsewhere.

Reply:
Sure, will make this change for all variables.